File: web_version/pmugg_files/network/primalEdge.js
// Interior is an vertex. Boundary should be null.
ms.primalEdge = function(type) {
	this.boundary = [];
	this.interior = [];
	this.type = type;

	this.id = ms.counter.add('primalEdge');
};

ms.primalEdge.prototype.getType     = function(){	return this.type; };
ms.primalEdge.prototype.getBoundary = function(){	return this.boundary[0]; };
ms.primalEdge.prototype.getInterior = function(){	return this.interior[0]; };

ms.primalEdge.prototype.export = function() {
	var i = this.getInterior();
	var type = i.getNetwork().getBoundNet().getTypes().edgeTypes.indexOf(this.type);
	return {
		interior: i ? i.getNetwork().edgeIndex(this.getInterior()) : -1,
		type,
	};
};

ms.primalEdge.prototype.import = function(boundNet, types, json){
	var interiorB = json.interior >= 0 ? boundNet.interior.getEdges()[json.interior] : null;
	ms.boundNet.connectInterior(this, interiorB);
	// var boundaryB = json.boundary >= 0 ? boundNet.boundary.getEdges()[json.boundary] : null;
	// ms.boundNet.connectBoundary(this, boundaryB);
	this.type = types.edgeTypes[json.type];
};

ms.primalEdge.prototype.copy = function(){
	return new ms.primalEdge(this.type);
};

ms.primalEdge.prototype.boundaryString = function() {
	return this.type.boundaryString();
};

ms.primalEdge.prototype.print = function() {
	this.interior.print();
};

ms.counter.register('primalEdge');

File: cpp_version/pmugg/network/primal_edge.cpp
#include "primal_edge.h"
#include "../shapes3D/edge_type3d.h"
#include "bound_net.h"
#include "../shapes3D/shape3d.h"
#include "../util/util.h"

namespace ms {

int PrimalEdge::nextId = 0;

PrimalEdge::PrimalEdge(EdgeType3D* edgeType)
    : type(edgeType)
    , id(nextId++) {}

std::string PrimalEdge::boundaryString() const {
    return type->boundaryString();
}

//void PrimalEdge::print() const {
//    if (!interior.empty()) {
//        interior[0]->print();
//    }
//}

PrimalEdge* PrimalEdge::copy() const {
    return new PrimalEdge(type);
}

void PrimalEdge::import(BoundNet* boundNet, Shape3D* shape, const Json& json) {
    // Get interior connection
    int interiorIndex = json["interior"].get<int>();
    if (interiorIndex >= 0) {
        auto* interiorB = boundNet->getInterior()->getEdges()[interiorIndex];
        BoundNet::connectInterior(this, interiorB);
    }

    // Set type
    type = shape->edgeTypes[json["type"].get<int>()];
}

//Json PrimalEdge::export() const {
//    Json json;
//    
//    // Export interior connection
//    auto* i = getInterior();
//    auto* network = i ? i->getNetwork() : nullptr;
//    json["interior"] = i ? network->edgeIndex(i) : -1;
//    
//    // Export type
//    json["type"] = type->getNetwork()->getBoundNet()->getTypes()->edgeTypes.indexOf(type);
//    
//    return json;
//}

} // namespace ms 